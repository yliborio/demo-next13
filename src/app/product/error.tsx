"use client";

export default function ErrorPage() {
  return <h1> Product not found </h1>;
}
