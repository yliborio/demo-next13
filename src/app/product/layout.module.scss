.container {
  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 1rem;
  margin: 1rem;
  .back {
    width: fit-content;
  }
}
