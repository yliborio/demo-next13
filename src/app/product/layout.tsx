import Link from "next/link";
import styles from "./layout.module.scss";
import { BackIcon } from "core/components/icons/back-button";
import { Suspense } from "react";
import Loading from "./loading";

export default function Layout({ children }: { children: React.ReactNode }) {
  return (
    <div className={styles["container"]}>
      <Link
        className={styles["back"]}
        href={"/"}
        aria-label="back to home page"
      >
        <BackIcon />
      </Link>
      <Suspense fallback={<Loading />}>{children}</Suspense>
    </div>
  );
}
