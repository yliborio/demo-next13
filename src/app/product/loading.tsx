import Skeleton from "react-loading-skeleton";
import "react-loading-skeleton/dist/skeleton.css";

export default function Loading() {
  return (
    <div
      style={{
        padding: "2rem 5rem",
      }}
    >
      <Skeleton
        height={"300px"}
        width={"100%"}
        count={1}
        baseColor="darkgrey"
      />
    </div>
  );
}
