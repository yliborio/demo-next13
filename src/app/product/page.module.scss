.container {
  display: flex;
  flex-direction: column;
  width: 30vw;
  align-items: center;
  justify-content: center;
  .title {
    font-size: 20px;
    line-height: 18px;
    margin-bottom: 1rem;
  }
  .action {
    display: flex;
    flex-direction: row;
    gap: 1rem;
    .input {
      height: 2rem;
    }
    .button {
      background-color: black;
      width: 25%;
    }
  }
}
