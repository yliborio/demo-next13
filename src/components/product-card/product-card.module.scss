@import "./../../styles/variables.scss";

.container {
  text-decoration: none;
  color: black;
  .card {
    display: flex;
    flex-direction: column;
    width: 100%;
    justify-content: center;
    align-items: center;
    gap: 1rem;
    padding: 1rem;
    background-color: $bg-secondary;
    border-radius: 8px;

    .title {
      display: flex;
      font-size: $font-size-2;
      height: 120px;
      justify-content: center;
      align-items: center;
    }

    .divider {
      width: 95%;
      height: 2px;
      background-color: darkgray;
    }

    .product-image {
      > img {
        width: 100%;
        height: 100%;
      }
      height: 200px;
    }
    .price {
      font-size: $font-size-2;
      font-weight: 500;
    }
  }
}
