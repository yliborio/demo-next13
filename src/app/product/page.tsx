import { ProductInfo } from "core/components/product-info/product-info";
import { FakeAPIProduct } from "core/types/product";

interface PageProps {
  searchParams: { id: string };
}

export default async function Page(props: PageProps) {
  const {
    searchParams: { id },
  } = props;
  const data = await fetch(`https://fakestoreapi.com/products/${id}`);
  const product: FakeAPIProduct = await data.json();

  return <ProductInfo product={product} />;
}
