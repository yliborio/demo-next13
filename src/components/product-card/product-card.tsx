import { FakeAPIProduct } from "core/types/product";
import styles from "./product-card.module.scss";
import Link from "next/link";
import Image from "next/image";

interface ProductCardProps {
  product: FakeAPIProduct;
}

export const ProductCard = (props: ProductCardProps) => {
  const { product } = props;

  return (
    <Link href={`/product?id=${product.id}`} className={styles["container"]}>
      <div className={styles["card"]}>
        <div className={styles["product-image"]}>
          <Image
            alt=""
            src={product.image}
            sizes="100vw"
            width={0}
            height={0}
          />
        </div>
        <div className={styles["title"]}>
          <span>{`${product.title} - ${product.rating.rate}/5  (${product.rating.count})`}</span>
        </div>
        <span className={styles["divider"]} />
        <span className={styles["price"]}>{`$${product.price}`} </span>
      </div>
    </Link>
  );
};
