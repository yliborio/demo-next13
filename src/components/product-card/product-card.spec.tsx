import React from "react";
import { render, screen } from "@testing-library/react";
import { ProductCard } from "./product-card";
import { FakeAPIProduct } from "core/types/product";

describe("ProductCard", () => {
  it("renders product information correctly", () => {
    const productData: FakeAPIProduct = {
      id: 1,
      title: "Sample Product",
      rating: { rate: 4.5, count: 100 },
      price: 19.99,
      image: "/sample-image-url",
      description: "",
      category: "",
    };

    const { asFragment } = render(<ProductCard product={productData} />);
    expect(asFragment()).toMatchSnapshot();

    expect(
      screen.getByText("Sample Product - 4.5/5 (100)")
    ).toBeInTheDocument();
    expect(screen.getByText("$19.99")).toBeInTheDocument();
  });
});
