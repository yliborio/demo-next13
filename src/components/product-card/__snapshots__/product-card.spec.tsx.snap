// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard renders product information correctly 1`] = `
<DocumentFragment>
  <a
    class="container"
    href="/product?id=1"
  >
    <div
      class="card"
    >
      <div
        class="product-image"
      >
        <img
          alt=""
          data-nimg="1"
          decoding="async"
          height="0"
          loading="lazy"
          sizes="100vw"
          src="/_next/image?url=%2Fsample-image-url&w=3840&q=75"
          srcset="/_next/image?url=%2Fsample-image-url&w=640&q=75 640w, /_next/image?url=%2Fsample-image-url&w=750&q=75 750w, /_next/image?url=%2Fsample-image-url&w=828&q=75 828w, /_next/image?url=%2Fsample-image-url&w=1080&q=75 1080w, /_next/image?url=%2Fsample-image-url&w=1200&q=75 1200w, /_next/image?url=%2Fsample-image-url&w=1920&q=75 1920w, /_next/image?url=%2Fsample-image-url&w=2048&q=75 2048w, /_next/image?url=%2Fsample-image-url&w=3840&q=75 3840w"
          style="color: transparent;"
          width="0"
        />
      </div>
      <div
        class="title"
      >
        <span>
          Sample Product - 4.5/5  (100)
        </span>
      </div>
      <span
        class="divider"
      />
      <span
        class="price"
      >
        $19.99 
      </span>
    </div>
  </a>
</DocumentFragment>
`;
